src/application-manager.coffee

```diff
@@ -0,0 +1,875 @@
+			else
+				@db.transaction(setInTransaction)
+		.then =>
+			@_targetVolatilePerServiceId = {}
+
+	setTargetVolatileForService: (serviceId, target) ->
+		@_targetVolatilePerServiceId[serviceId] ?= {}
+		_.assign(@_targetVolatilePerServiceId, target)
+
+	getTargetApps: =>
+		Promise.map(@db.models('app').select(), @normaliseAndExtendAppFromDB)
+		.map (app) =>
+			if !_.isEmpty(app.services)
+				app.services = _.map app.services, (service) =>
+					_.merge(service, @_targetVolatilePerServiceId[service.serviceId]) if @_targetVolatilePerServiceId[service.serviceId]?
+					return service
+			return app
+
+	getDependentTargets: =>
+		@proxyvisor.getTarget()
+
+	bestDeltaSource: (image, available) ->
+		if !image.dependent
+			for availableImage in available
+				if availableImage.serviceName == image.serviceName and availableImage.appId == image.appId
+					return availableImage.name
+			for availableImage in available
+				if availableImage.serviceName == image.serviceName
+					return availableImage.name
+		for availableImage in available
+			if availableImage.appId == image.appId
+				return availableImage.name
+		return 'resin/scratch'
+
+	# return images that:
+	# - are not used in the current state, and
+	# - are not going to be used in the target state, and
+	# - are not needed for delta source / pull caching or would be used for a service with delete-then-download as strategy
+	_unnecessaryImages: (current, target, available) =>
+
+		allImagesForApp = (app) -> _.map(app.services, imageForService)
+
+		currentImages = _.flatten(_.map(current.local?.apps, allImagesForApp))
+		targetImages = _.flatten(_.map(target.local?.apps, allImagesForApp))
+		availableAndUnused = _.filter available, (image) ->
+			!_.some currentImages.concat(targetImages), (imageInUse) -> image.name == imageInUse.name
+		imagesToDownload = _.filter targetImages, (imageName) ->
+			!_.some available, (availableImage) -> availableImage.name == imageName.name
+
+		deltaSources = _.map imagesToDownload, (image) =>
+			return @bestDeltaSource(image, available)
+
+		proxyvisorImages = @proxyvisor.imagesInUse(current, target)
+
+		return _.filter availableAndUnused, (image) ->
+			notUsedForDelta = !_.some deltaSources, (deltaSource) -> deltaSource == image.name
+			notUsedByProxyvisor = !_.some proxyvisorImages, (proxyvisorImage) -> image.name == proxyvisorImage
+			return notUsedForDelta and notUsedByProxyvisor
+
+	_inferNextSteps: (cleanupNeeded, availableImages, current, target, stepsInProgress) =>
+		Promise.try =>
+			currentByAppId = _.keyBy(current.local.apps ? [], 'appId')
+			targetByAppId = _.keyBy(target.local.apps ? [], 'appId')
+			nextSteps = []
+			if !_.some(stepsInProgress, (step) -> step.action == 'fetch')
+				if cleanupNeeded
+					nextSteps.push({ action: 'cleanup' })
+				imagesToRemove = @_unnecessaryImages(current, target, availableImages)
+				_.forEach imagesToRemove, (image) ->
+					nextSteps.push({ action: 'removeImage', image })
```

Comment:
when the first image is removed, `_inferNextSteps` will re-run and the image removal steps will be re-added

Reply:
Fixed by making the `@_removeDuplicateSteps` also affect these steps